src/python/pants/backend/go/pkg.py

```diff
@@ -0,0 +1,256 @@
+    # execution sandbox. Thus, this code uses a bash script to be able to resolve that path.
+    # TODO: Merge all duplicate versions of this script into a single script and add an invoke rule that will
+    # insert the desired `go` command into the boilerplate portions.
+    analyze_script_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(
+                    "analyze.sh",
+                    textwrap.dedent(
+                        f"""\
+                export GOROOT="$(/bin/pwd)/go"
+                export GOPATH="$(/bin/pwd)/gopath"
+                export GOCACHE="$(/bin/pwd)/cache"
+                /bin/mkdir -p "$GOPATH" "$GOCACHE"
+                cd {resolved_go_module.address.spec_path}
+                exec "${{GOROOT}}/bin/go" list -json ./{spec_subpath}
+                """
+                    ).encode("utf-8"),
+                )
+            ]
+        ),
+    )
+
+    input_root_digest = await Get(
+        Digest,
+        MergeDigests([sources.snapshot.digest, downloaded_goroot.digest, analyze_script_digest]),
+    )
+
+    process = Process(
+        argv=[bash.path, "./analyze.sh"],
+        input_digest=input_root_digest,
+        description="Resolve go_package metadata.",
+        level=LogLevel.DEBUG,
+    )
+
+    result = await Get(ProcessResult, Process, process)
+
+    metadata = json.loads(result.stdout)
+
+    # TODO: Raise an exception on errors. They are only emitted as warnings for now because the `go` tool is
+    # flagging missing first-party code as a dependency error. But we want dependency inference and won't know
+    # what the dependency actually is unless we first resolve the package with that dependency. So circular
+    # reasoning. We may need to hydrate the sources for all go_package targets that share a `go_module`.
+    if metadata.get("Incomplete"):
+        error_dict = metadata.get("Error", {})
+        if error_dict:
+            error_str = error_to_string(error_dict)
+            _logger.warning(
+                f"Error while resolving Go package at address {request.address}: {error_str}"
+            )
+        # TODO: Check DepsErrors key as well.
+
+    # Raise an exception if any unsupported source file keys are present in the metadata.
+    for key in (
+        "CompiledGoFiles",
+        "CFiles",
+        "CXXFiles",
+        "MFiles",
+        "HFiles",
+        "FFiles",
+        "SFiles",
+        "SwigFiles",
+        "SwigCXXFiles",
+        "SysoFiles",
+    ):
+        files = metadata.get(key, [])
+        if files:
+            raise ValueError(
+                f"The go_package at address {request.address} contains the following unsupported source files "
+                f"that were detected under the key '{key}': {', '.join(files)}."
+            )
+
+    package_name: str = metadata["Name"]
+    imports = typing.cast(Tuple[str, ...], tuple(metadata.get("Imports", [])))
```

Comment:
We normally directly import `cast` from `typing`. Not blocking, but might make this code slightly more readable.

Reply:
Or, probably even better is that you can inline these all into the `ResolvedGoPackage` constructor and simply use `# type: ignore[whatever-the-error-code-is]`. That's functionally equivalent here to using `cast`, and slightly less boilerplatey because you don't need to declare each to be `Tuple[str, ...]`.

Reply:
I've moved these into the dataclass, and mypy is not even requiring a type ignore.